File: sudoku.lua
-- Sudoku solving using backtrack
-- Unit testing starts

ROW_SIZE = 9

-- progress bar
function load_bar(step, total_steps, resolution, width)
   if math.floor(total_steps/resolution) == 0 then
      return
   end
   if step % math.floor(total_steps/resolution) ~= 0 then
      return
   end
   ratio = step/total_steps
   count = math.floor(ratio * width)
   io.write(string.format("%3d%% [", math.floor(ratio * 100)))
   for i=1, count do
      io.write("=")
   end
   for i=count, width-1 do
      io.write(" ")
   end
   io.write("]\r")
   io.flush()
end


-- split a string by a separator
function string:split(sep)
	local sep, fields = sep or ":", {}
	local pattern = string.format("([^%s]+)", sep)
	self:gsub(pattern, function(c) fields[#fields+1] = c end)
	return fields
end


-- get a 2D array sudoku from a 81 character string
function from_str(text)
	local board = {}
	for i=1, #text, ROW_SIZE do
		local row = {}
		for c in string.gmatch(text:sub(i, i+ROW_SIZE-1), ".") do
			table.insert(row, tonumber(c))
		end
		table.insert(board, row)
	end
	return board
end


-- pretty print a sudoku
function to_str(sudoku)
	local board = ""
	for i=1, #sudoku do
		for j=1, #sudoku[i] do
			board = board .. sudoku[i][j] .. " "
		end
		board = board .. "\n"
	end
	board = board .. "\n"
	return board
end

function next_empty(a_sudoku)
	for i=1, #a_sudoku do
		for j=1, #a_sudoku[i] do
			if a_sudoku[i][j] == 0 then
				return {i, j}
			end
		end
	end
	return nil
end


function can_put(a_sudoku, row, col, val)
	for x=1, ROW_SIZE do
		if a_sudoku[row][x] == val then return false end -- test row
		if a_sudoku[x][col] == val then return false end -- test column
	end
        local row_ = row-1
        local col_ = col-1
	local a, b = row_-(row_%3), col_-(col_%3)
	for i=a, a+2 do
		for j=b, b+2 do
			if a_sudoku[i+1][j+1] == val then return false end -- test square
		end
	end
	return true
end

function solve(sudoku)

	local spot = next_empty(sudoku)
	if spot then
		local x, y = spot[1], spot[2]
		for val=1, 9 do
			if can_put(sudoku, x, y, val) then
				sudoku[x][y] = val
				local new_sudoku = solve(sudoku)
				if next_empty(new_sudoku) == nil then
					-- solution found
					return new_sudoku
				end
			end
		end
		-- solution not found, backtrack
		sudoku[x][y] = 0
	end
	return sudoku
end


-- Read sudokus into memory
inputname = arg[1] or 'input_small.txt'
f = io.input(inputname)
data = io.read('*a')
f:close()

-- Solve all sudokus
before = os.time()
lines = data:split("\n")
total = #lines
solved = 0
result = ''
for i=1, #lines do
	result = result .. to_str(solve(from_str(lines[i])))
    solved = solved + 1
    load_bar(solved, total, 20, 50)
end

-- present time
print('-- Elapsed time ' .. (os.time() - before) .. ' s')

-- write solutions to file
f = io.output('solved_' .. inputname)
f:write(result)
f:close()

File: sudoku.pl
# solve a sudoku using backtrack
use strict;
use warnings;
use Scalar::Util qw( looks_like_number );
use Time::HiRes qw( time );
use constant ROW_SIZE => 9;

sub solve {

    sub next_empty {
        my $sudoku = shift;
        for my $row (0 .. ROW_SIZE-1) {
            for my $col (0 .. ROW_SIZE-1) {

                if ($$sudoku[$row][$col] == 0) { return ($row, $col); }
            }
        }
        return ();
    }

    sub can_put {
        my ($sudoku, $x, $y, $val) = @_;
        for my $i (0 .. ROW_SIZE - 1) {
            if ($$sudoku[$x][$i] == $val) { return 0; }
            if ($$sudoku[$i][$y] == $val) { return 0; }
        }
        my $sq_x = $x - ($x % 3);
        my $sq_y = $y - ($y % 3);
        for my $row ($sq_x .. $sq_x + 2) {
            for my $col ($sq_y .. $sq_y + 2) {
                if ($$sudoku[$row][$col] == $val) {
                    return 0;
                }
            }
        }
        return 1;

    }

    my $sudoku = shift;
    my @spot = next_empty $sudoku;
    if (@spot) {
        my ($x, $y) = @spot;
        for my $val (1 .. 9) {
            if (can_put $sudoku, $x, $y, $val) {
                $$sudoku[$x][$y] = $val;
                my $new_sudoku = solve ($sudoku);
                my @new_spot = next_empty $new_sudoku;
                if (@new_spot == 0) {return $new_sudoku; } # solution found
            }
        }
        $$sudoku[$x][$y] = 0; # backtrack
    }
    return $sudoku;
}

sub from_str {
    my @sudoku = ();
    my $str = shift;
    my @numbers = grep { looks_like_number($_) } split(//, $str);
    for my $i (map {ROW_SIZE * $_} 0..ROW_SIZE-1) {
        push @sudoku, [@numbers[$i .. ($i+ROW_SIZE-1)]];
    }
    return @sudoku;
}

sub to_str {
    my $sudoku = shift;
    my $result = "";
    foreach my $row (@$sudoku) {
        $result .= " @$row \n";
    }

    return $result;
}

sub load_bar {
    my ($step, $totalSteps, $resolution, $width) = @_;
    if (int($totalSteps/$resolution) != 0 and 
        $step % int($totalSteps/$resolution) == 0) {
        my $ratio = $step/$totalSteps;
        my $count = int($ratio * $width);
        printf("%d%% [",int($ratio * 100));
        for (0 .. $count) { printf("="); }
        for ($count+1 .. $width) { printf(" "); }
        printf("]\r");
        $|++; #flush
    }
}

# read sudokus into memory
my $input = $ARGV[0];
open my $in, "<", $input or die "Can't open input file";
my $data;
read $in, $data, -s $input;
close $in;

my $result = "";
my $start_time = time();

# solve all sudokus
my @rows = split '\n', $data;
my $total = @rows;
for my $step (0 .. $#rows) {
    my @sudoku = from_str($rows[$step]);
    $result .= to_str( solve (\@sudoku)) . "\n";
    load_bar($step, $total, 20, 50);
}
my $diff = time() - $start_time;

# write solutions to file
open my $out, ">", "solved_" . $input or die "Can't open output file";
print $out $result;
close $out;

print " -- Elapsed time: $diff s.\n";
